### PR TITLE
[TRTLLM-16012][feat] Restrict torch.compile to piecewise graph capture only

### PR DESCRIPTION
**Pull request by AlessioNetti**
## Dev Engineer Review

- Added `TorchCompileConfig.compile_only_piecewise_graphs`, defaulting to `False`.
- Added selective eager and compiled forward dispatch.
- Bypasses `torch.compile` for generation-only batches and auxiliary warmup operations.
- Preserves compiled execution for piecewise CUDA graph capture and warmup.
- Added validation for unsupported model types.
- Updated the golden manifest with the new configuration field.
- Reported startup time improvements for Qwen3 8B FP8.

## QA Engineer Review

- Added unit coverage for phase-based dispatch and bypass restoration.
- Added `TestQwen3_0_6B.test_piecewise_cuda_graph_compile_restriction_stability()`.
- The integration test compares generated tokens and logits between restricted and unrestricted compilation.
- The integration test is covered in:
  - `tests/integration/test_lists/qa/llm_function_core.txt`
  - `tests/integration/test_lists/test-db/l0_h100.yml`
- **Verdict: sufficient.**

## Description

Enabling `torch.compile` in TRT-LLM via `TorchCompileConfig`, for piecewise CUDA graphs, often results in a large increase in startup times. This is because, when this feature is enabled, the entire model forward is wrapped unconditionally in `torch.compile` at the beginning of the initialization process: this means that compilation is invoked redundantly during many parts of the initialization process (e.g., attention warmup, auto-tuning) and for input shapes which are not needed during inference.

This PR addresses the issue above by introducing a new `compile_only_piecewise_graphs` knob to `TorchCompileConfig`: this prevents invocation of `torch.compile` in the following instances:

1) For generation-only batches, where ordinary CUDA graphs come into play;
2) For all initialization operations such as warmup and auto-tuning.

In practice, this means that `torch.compile` is invoked only for piecewise CUDA graph capture and warmup. This leads to much faster startup times, see the table below for a comparison of the overall initialization time with a Qwen3 8B model (FP8 block scaling) on H100 NVL:

| Config  | torch.compile disabled | torch.compile enabled | torch.compile restricted |
| ------------- | ------------- | ------------- | ------------- |
| TP1 | ~53s  | ~115s | ~93s |
| TP2 | ~80s  | ~143s | ~96s |

The TP2 configuration sees a larger speedup because it involves more operations that are affected by the `torch.compile` overhead (in this case, allreduce auto-tuning).

At the same time, the effect on performance appears to be negligible - see the Pareto plot below, obtained with 15 ISL, 10 OSL (a regime where piecewise CUDA graphs are often needed), with concurrency levels from 1 to 256, at TP1.

plot_8b_15_10_tp1

Below is instead the same performance plot, obtained with TP2.

plot_8b_15_10_tp2_alt

## Test Coverage

New tests covering the torch.compile bypass mechanism have been added to `tests/unittest/_torch/compilation/test_phase_selective_forward.py`.

### CODE REVIEW

**Review by tburt-nv — APPROVED** [review]

File: tensorrt_llm/_torch/compilation/utils.py [hunk 1]
```diff
@@ -1,11 +1,60 @@
...
+        2) The batch is generation-only (rely on ordinary CUDA graphs)
+        """
+        attrs = get_model_extra_attrs()
+        attn_metadata_ref = attrs.get("attention_metadata") if attrs else None
+        attn_metadata = (attn_metadata_ref()
+                         if attn_metadata_ref is not None else None)
+        # This is the finalized batch classification after any
+        # context-to-generation promotion.
+        if (self._bypass_active or
+            (attn_metadata is not None and attn_metadata.num_contexts == 0)):
```
Comment (chienchunhung) [inline tensorrt_llm/_torch/compilation/utils.py]: This dispatch must be rank-uniform and tied to actual CUDA-graph eligibility. `attention_metadata.num_contexts` is rank-local under Attention DP; one rank can be generation-only while another has context work, so this can send collective-bearing decoder bodies through eager and the compiled/fused backend on different ranks. It also forces eager when decode CUDA-graph lookup returns no key, even though the comment assumes an ordinary CUDA graph will run. 

Suggestion: 
- Propagate the group-uniform graph/capture decision into this wrapper, or, 
- Reject Attention DP and graph-fallback configurations, keep compiled execution for generation fallback. 
- Add mixed-phase ADP plus no-key regressions.

File: tensorrt_llm/llmapi/llm_args.py [hunk 2]
```diff
@@ -5233,6 +5233,14 @@ class TorchCompileConfig(StrictBaseModel):
         default=True,
         description="Enable full graph compilation in torch.compile.")
 
+    compile_only_piecewise_graphs: bool = Field(
+        default=False,
+        description="Compile only context and mixed-batch execution graphs. "
+        "When enabled, generation-only and auxiliary kernel warmup forwards "
+        "use eager execution, while dedicated context/mixed compilation and "
+        "piecewise CUDA graph warmup and capture remain compiled.",
+        status="prototype")
```
Comment (chienchunhung) [inline tensorrt_llm/llmapi/llm_args.py:5380]: `compile_only_piecewise_graphs=True` is currently accepted while `prefill_cuda_graph_backend` remains `disabled`. In that configuration the backend is not piecewise, but every context/mixed forward still enters `torch.compile`, so the option silently violates its name and PR contract. 

We should either require `PrefillCudaGraphBackend.PIECEWISE` in `normalize_prefill_cuda_graph_config()` and dispatch on actual piecewise eligibility, or rename the option to describe phase-selective compilation. 

We should also add validation tests for disabled and conflicting backends.

**Review by chienchunhung — COMMENTED** [review]
Thanks for the PR!

**Review by chienchunhung — CHANGES_REQUESTED** [review]
Summarizing my overall observations:

The startup optimization is worthwhile: compiling generation-only forwards can be unnecessary when execution will reliably use ordinary CUDA-graph replay, while context/mixed execution still needs `torch.compile` for piecewise graphs and compiler transformations. 

However, the current rank-local `num_contexts` check does not establish actual CUDA-graph eligibility: graph misses will unexpectedly use eager execution, potentially regressing performance, and Attention-DP ranks may choose different eager/compiled collective paths, creating a correctness risk. 

**Suggestion:** Please move the decision into `ModelEngine`, derive it from the actual group-uniform graph mode, and define an explicit graph-miss fallback. The option should also require piecewise prefill—or be renamed to describe phase-selective compilation—and include numerical, performance, graph-miss, and mixed-phase multi-rank tests. With those constraints, this is a valid and useful opt-in optimization.

**Review by nvxuanyuc — COMMENTED** [review]
@AlessioNetti I have a few concerns about this change:
1. Does retaining both eager and compiled forward paths introduce additional memory overhead? Do we have before-and-after device-memory measurements?
2. Eager and compiled execution may produce numerical differences. Since this change switches between the two modes within the same request lifecycle, do we have stability tests covering these transitions?
3. torch.compile can still provide meaningful kernel-fusion benefits even when generation-only batches use full cuda graphs. Have we verified that capturing the eager generation path does not cause a performance regression? Speculative decoding may be particularly sensitive to this change.
Appreciate any clarification or additional validation you can share on these points. Thanks!

**Comment by AlessioNetti** [comment]
> @AlessioNetti I have a few concerns about this change:
> 
>     1. Does retaining both eager and compiled forward paths introduce additional memory overhead? Do we have before-and-after device-memory measurements?
> 
>     2. Eager and compiled execution may produce numerical differences. Since this change switches between the two modes within the same request lifecycle, do we have stability tests covering these transitions?
> 
>     3. torch.compile can still provide meaningful kernel-fusion benefits even when generation-only batches use full cuda graphs. Have we verified that capturing the eager generation path does not cause a performance regression? Speculative decoding may be particularly sensitive to this change.
>        Appreciate any clarification or additional validation you can share on these points. Thanks!

Thanks for the feedback @nvxuanyuc - here's some context:

1. The main memory overhead comes from enabling `torch.compile` via `TorchCompileConfig`. According to TRT-LLM's logged memory profiling, doing this lifts peak `torch + non-torch` memory usage from 13.38GB to 13.63GB, a ~250MB increase (Qwen3 8B FP8, TP1, H100 NVL). Additionally setting `compile_only_piecewise_graphs=True` does not seem to affect this further.
2. Yes, this is true. I'll look into adding a test for output stability - in general, since this is an opt-in optimization, I would think that potential numerical differences (as well as perf regressions, see below) should be tolerable as long as this is documented properly.
3. The target use case for this PR is a fairly specific one, i.e., small model + very short OSL. In this case, capturing piecewise CUDA graphs is fundamental to prevent most iterations from falling back to eager mode, while other benefits coming from `torch.compile` are not that relevant; this is also shown in the performance comparison at the top of the PR, where setting `compile_only_piecewise_graphs=True` does not hurt performance. In other instances (e.g., spec dec, like you said) there might indeed be performance regressions, in which case we should advise to keep `compile_only_piecewise_graphs` disabled.

Hope this addresses your concerns!

**EDIT**: Following up on point 2, I added a simple output stability test with commit 6d045450f6b30f8a357f2244f78749aa2e1c5796, and added a disclaimer about potential perf degradations (compared to full `torch.compile`) with commit 165254cfb5dda190eedd5c258e530f00fc915cb7.

File: tensorrt_llm/llmapi/llm_args.py (hunk 2 above)
Comment (AlessioNetti) [reply tensorrt_llm/llmapi/llm_args.py:5380]: Thanks - I had missed this. I added a check during config validation with commit 39249ef4724125a2de2b32ed8553874591ffabe2 to prevent this, and also forbid using the new option under attention DP (see other comment). I also added simple validation tests with commit 08f5a046188c0a8c09fc6c49339e67560737eee5.

File: tensorrt_llm/_torch/compilation/utils.py (hunk 1 above)
Comment (AlessioNetti) [reply tensorrt_llm/_torch/compilation/utils.py]: For now I opted to simply forbid using `compile_only_piecewise_graphs` together with attention DP in commit 39249ef4724125a2de2b32ed8553874591ffabe2 - we had a similar problem with PR [17412](https://github.com/NVIDIA/TensorRT-LLM/pull/17412) (CUDA graph specialization for LoRA), where in the end we decided that the extra complexity was not worth adding unless there is a strong use case - of course, we can revisit this in the future.

**Comment by AlessioNetti** [comment]
> Suggestion: Please move the decision into ModelEngine, derive it from the actual group-uniform graph mode, and define an explicit graph-miss fallback. The option should also require piecewise prefill—or be renamed to describe phase-selective compilation—and include numerical, performance, graph-miss, and mixed-phase multi-rank tests. With those constraints, this is a valid and useful opt-in optimization.

@chienchunhung I moved the compile/eager selection logic inside of `PyTorchModelEngine` as suggested in commit cf28505af7680eac725f78e86118a822bf65a885, and I am now using existing graph eligibility flags. This should prevent awkward edge cases such as a non-graph-eligible prefill batch (e.g. due to large token count) triggering `torch.compile`.

File: tests/integration/defs/accuracy/test_llm_api_pytorch.py [hunk 3]
```diff
@@ -4688,6 +4688,56 @@ def test_nvfp4(self, ep_size, attention_dp):
...
+                output = llm.generate(
+                    prompts, sampling_params=sampling_params)[0].outputs[0]
+                assert output.generation_logits is not None
+                return list(output.token_ids), output.generation_logits.clone()
+
+        compiled_tokens, compiled_logits = run(False)
+        restricted_tokens, restricted_logits = run(True)
+
+        assert restricted_tokens == compiled_tokens
+        torch.testing.assert_close(restricted_logits,
```
Comment (crazydemo) [inline tests/integration/defs/accuracy/test_llm_api_pytorch.py]: I feel like this is more like a unit test, in accuracy test suite, we usually do accuracy benchmark (e.g. mmlu), so that we can keep tracking the accuracy regression.

Comment (AlessioNetti) [reply tests/integration/defs/accuracy/test_llm_api_pytorch.py]: Hi - your assessment is correct, this is mostly meant to be a unit test to ensure basic output consistency when `compile_only_piecewise_graphs` is enabled. This is a fairly specific, opt-in startup optimization, so implementing an extensive accuracy suite shouldn't be necessary. Should I move the test under the `unittest` directory?

**Review by chienchunhung — APPROVED** [review]
Thanks for addressing my previous comments. Overall LGTM. 

The parity test verifies one output comparison, but it does not prove that restricted mode actually selects the eager generation path; it would still pass if both configurations used the compiled decoder, and neither documented graph-miss branch is exercised. Since the contract now depends on `can_run_graph` and `get_per_request_prefill_cuda_graph_flag()`, please add engine-level behavioral coverage for:
- PCG-eligible prefill → compiled, 
- PCG-ineligible prefill → eager, 
- ordinary generation capture → eager, and 
- generation graph miss → eager. 

The reported performance comparison is encouraging for the targeted short-OSL TP1 workload, but it does not establish that capturing eager generation forwards is **_generally_** equivalent to capturing compiled forwards. Let's document the validated workload explicitly and call out unvalidated cases such as longer generation, TP2+, speculative decoding, and generation graph misses. Since the option is prototype and disabled by default, broader benchmarking need not necessarily block this PR, but it should be required before enabling the optimization automatically or presenting it as generally performance-neutral.

**Review by Mgluhovskoi — APPROVED** [review]

**Review by arysef — APPROVED** [review]

**Review by mzweilz — APPROVED** [review]

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 4]
```diff
@@ -1367,6 +1387,30 @@ def _pad_batch_seed_mrope_delta_cache(
         for request in mrope_seed_requests:
             request.py_mrope_delta_cache_slot = request.py_seq_slot
 
+    @contextmanager
+    def _maybe_bypass_torch_compile(self,
+                                    bypass: bool = False,
+                                    can_run_graph: bool = True):
```
Comment (cascade812) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py]: `can_run_graph: bool = True` makes the default behaviour of a bare `self._maybe_bypass_torch_compile()` call *bypass compilation*. That's the wrong failure direction for a default — a future call site that forgets the argument silently drops into eager for every forward, and nothing will fail loudly. Every current caller passes it explicitly, so make it keyword-only and required (or default it to `False`).

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 5]
```diff
@@ -1552,14 +1599,16 @@ def warmup(self, resource_manager: ResourceManager) -> None:
         # Helix CP is decode-only and runs into issues with the
         # autotuner warmup's context requests.
         if not is_enc_dec and not self.mapping.has_cp_helix():
-            self._run_autotuner_warmup(resource_manager)
+            with self._maybe_bypass_torch_compile(bypass=True):
+                self._run_autotuner_warmup(resource_manager)
```
Comment (cascade812) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py]: Bypassing the autotuner warmup means tactic exploration runs against the uncompiled decoder. Ops that only exist after the compile backend's passes (userbuffer allreduce, fused rmsnorm/quant variants) are never profiled, so when the piecewise graph later invokes them the autotuner is out of exploration mode and falls back to the default tactic. Worth confirming against the autotuner's fallback warnings with `enable_userbuffers=True` and `enable_inductor=True`, and noting the caveat in the docs section alongside the existing "evaluate case-by-case" line.

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 6]
```diff
@@ -1603,10 +1652,12 @@ def warmup(self, resource_manager: ResourceManager) -> None:
         log_mem_snapshot("warmup/after_cute_dsl_radix_topk")
         if can_run_general_warmup:
             # Pre-populate the memory pool with max-shape allocations to reduce
-            # fragmentation at runtime.
+            # fragmentation at runtime. If compile_only_piecewise_graphs is
```
Comment (cascade812) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py]: The comment asserts the bypass is "safe" here, but this warmup exists to pre-populate the caching allocator at max shape so runtime doesn't fragment. The compiled graph's intermediate buffers (inductor-allocated temporaries, aux-stream buffers) don't necessarily match the eager forward's allocation profile, so the pool may be sized against the wrong peak. Either justify why the compiled path allocates no more than eager, or leave this one compiled — it runs once and is not where the startup cost is.

File: tensorrt_llm/_torch/compilation/utils.py [hunk 7]
```diff
@@ -8,6 +12,39 @@
 from ..cuda_tile_utils import IS_CUDA_TILE_AVAILABLE
 
 
+class _PhaseSelectiveForward:
```
Comment (cascade812) [inline tensorrt_llm/_torch/compilation/utils.py]: `_PhaseSelectiveForward` is underscore-private but imported across packages by `pyexecutor/model_engine.py` and by the new test. Drop the underscore (`PhaseSelectiveForward`) so the cross-module dependency is honest, or keep it private and expose a factory.

File: tests/unittest/_torch/compilation/test_phase_selective_forward.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+    assert forward() == "compiled"
+
+
+@pytest.mark.parametrize(
+    ("can_run_graph", "prefill_graph_eligible", "expected"),
+    [
+        pytest.param(False, True, "compiled", id="pcg-eligible-prefill"),
+        pytest.param(False, False, "eager", id="pcg-ineligible-prefill"),
+        pytest.param(True, True, "eager", id="ordinary-generation-capture"),
+        pytest.param(False, False, "eager", id="generation-graph-miss"),
```
Comment (cascade812) [inline tests/unittest/_torch/compilation/test_phase_selective_forward.py]: `generation-graph-miss` is `(False, False)` — byte-identical inputs to `pcg-ineligible-prefill` on line 42, just a different id. The case you actually want and are missing is `(can_run_graph=True, prefill_graph_eligible=False)`, which asserts that an ordinary-graph hit short-circuits before the eligibility flag is read. Also add a case with `engine._phase_selective_forward = None` to pin the no-op path, since that's what every non-`compile_only_piecewise_graphs` run takes.

File: tests/unittest/_torch/compilation/test_phase_selective_forward.py [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+
+    assert actual == "eager"
+    eager_forward.assert_called_once_with()
+    compiled_forward.assert_not_called()
+    get_prefill_graph_flag.assert_not_called()
+
+
+@skip_pre_hopper
+@pytest.mark.timeout(1800)
+def test_piecewise_cuda_graph_compile_restriction_stability() -> None:
```
Comment (cascade812) [inline tests/unittest/_torch/compilation/test_phase_selective_forward.py]: This builds two full `LLM` instances with `torch.compile` and a 30-minute timeout inside `tests/unittest/_torch/compilation/`, a directory listed wholesale in both `l0_h100.yml` and `l0_b200.yml` — so it lands in the unit-test stages on two GPU configs. Move it to `tests/integration/defs/` and register it explicitly, keeping this file to the mock-based dispatch tests. Separately, the two `pytest.skip` guards mean the test silently disappears if `LLM_MODELS_ROOT` or the Qwen3 path moves; `llm_models_root(check=True)` would fail loudly instead.

**Comment by chienchunhung** [comment]
@AlessioNetti FYI there are merge conflicts for this PR. Please rebase and trigger the CI again. Thanks!
